common/codex32.c

```diff
@@ -13,6 +13,186 @@
+	}
+}
+
+/* Helper to input the HRP of codex32 string into the checksum engine */
+static void input_hrp(struct checksum_engine *codex_checksum_engine, const char *hrp)
+{
+	size_t i = 0;
+	for (i = 0; i < strlen(hrp); i++) {
+		input_fe(codex_checksum_engine, tolower(hrp[i]) >> 5);
+	}
+	input_fe(codex_checksum_engine, tolower(hrp[i]) >> 0);
+	for (i = 0; i < strlen(hrp); i++) {
+		input_fe(codex_checksum_engine, tolower(hrp[i]) & 0x1f);
+	}
+	return;
+}
+
+/* Helper to input data strong of codex32 into the checksum engine. */
+static void input_data_str(struct checksum_engine *codex_checksum_engine, const char *datastr)
+{
+	size_t i = 0;
+
+	for (i = 0; i < strlen(datastr); i++) {
+		input_fe(codex_checksum_engine, bech32_charset_rev[(int)datastr[i]]);
+	}
+
+	return;
+}
+
+/* Helper to verify codex32 checksum */
+static bool checksum_verify_short (const char *hrp, const char *codex_datastr)
+{
+	struct checksum_engine *codex_checksum_engine = tal(tmpctx, struct checksum_engine);
+
+	uint8_t generator[13] = {
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'M'],
+		bech32_charset_rev[(int)'3'], bech32_charset_rev[(int)'G'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'E'],
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'E'],
+		bech32_charset_rev[(int)'L'], bech32_charset_rev[(int)'M'],
+		bech32_charset_rev[(int)'C'], bech32_charset_rev[(int)'S'],
+		bech32_charset_rev[(int)'S']
+	};
+	uint8_t residue[13] = {
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'Q'], bech32_charset_rev[(int)'Q'],
+		bech32_charset_rev[(int)'P']
+	};
+	uint8_t target[13] = {
+		bech32_charset_rev[(int)'S'], bech32_charset_rev[(int)'E'],
+		bech32_charset_rev[(int)'C'], bech32_charset_rev[(int)'R'],
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'T'],
+		bech32_charset_rev[(int)'S'], bech32_charset_rev[(int)'H'],
+		bech32_charset_rev[(int)'A'], bech32_charset_rev[(int)'R'],
+		bech32_charset_rev[(int)'E'], bech32_charset_rev[(int)'3'],
+		bech32_charset_rev[(int)'2']
+	};
+	codex_checksum_engine->generator = tal_dup_arr(codex_checksum_engine, u8, generator, 13, 0);
+	codex_checksum_engine->residue = tal_dup_arr(codex_checksum_engine, u8, residue, 13, 0);
+	codex_checksum_engine->target = tal_dup_arr(codex_checksum_engine, u8, target, 13, 0);
+
+	input_hrp(codex_checksum_engine, hrp);
+	input_data_str(codex_checksum_engine, codex_datastr);
+
+	if (!streq(tal_hexstr(tmpctx,
+			      codex_checksum_engine->target,
+			      tal_bytelen(codex_checksum_engine->target)),
+		     tal_hexstr(tmpctx,
+		        	codex_checksum_engine->residue,
+				tal_bytelen(codex_checksum_engine->residue)))) {
+			return false;
```

Comment:
This is a very complex way of doing "memeq(target, sizeof(target), residue, sizeof(residue))"!